presto-main/src/test/java/com/facebook/presto/split/TestBufferingSplitSource.java

```diff
@@ -13,89 +13,214 @@
+
+            // Source failed after getNextBatch invocation.
+            nextBatchFuture = invokeNextBatch(source, 10, nextBatchCall);
+            mockSource.increaseAvailableSplits(5);
+            assertFalse(nextBatchFuture.isDone());
+            mockSource.atSplitCompletion(FAIL);
+            assertNextBatchFailure(nextBatchFuture, nextBatchCall);
+            assertFalse(source.isFinished());
+        }
+
+        // Fast source: source produce 8 before, and 8 after invocation. BufferedSource should return all 16 at once.
+        mockSource = new MockSplitSource().setBatchSize(8);
+        try (SplitSource source = new BufferingSplitSource(mockSource, 10)) {
+            mockSource.increaseAvailableSplits(8);
+            ListenableFuture<?> nextBatchFuture = invokeNextBatch(source, 20, nextBatchCall);
+            assertFalse(nextBatchFuture.isDone());
+            mockSource.increaseAvailableSplits(8);
+            assertNextBatchResult(nextBatchFuture, 16, false, nextBatchCall);
         }
     }
 
-    @Test
-    public void testFail()
+    @Test(dataProvider = "nextBatchCalls")
+    public void testFailImmediate(NextBatchCall nextBatchCall)
             throws Exception
     {
-        MockSplitSource mockSource = new MockSplitSource(1, 2, 1);
+        MockSplitSource mockSource = new MockSplitSource().setBatchSize(1).atSplitCompletion(FAIL);
         try (SplitSource source = new BufferingSplitSource(mockSource, 100)) {
-            try {
-                source.getNextBatch(2).get();
-                fail();
-            }
-            catch (ExecutionException e) {
-                assertTrue(e.getCause() instanceof IllegalStateException);
-                assertEquals(e.getCause().getMessage(), "Mock failure");
+            assertNextBatchFailure(invokeNextBatch(source, 200, nextBatchCall), nextBatchCall);
+            assertEquals(mockSource.getNextBatchInvocationCount(), 1);
+        }
+    }
+
+    @Test(dataProvider = "nextBatchCalls")
+    public void testFail(NextBatchCall nextBatchCall)
+            throws Exception
+    {
+        MockSplitSource mockSource = new MockSplitSource().setBatchSize(1).increaseAvailableSplits(1).atSplitCompletion(FAIL);
+        try (SplitSource source = new BufferingSplitSource(mockSource, 100)) {
+            assertNextBatchFailure(invokeNextBatch(source, 2, nextBatchCall), nextBatchCall);
+            assertEquals(mockSource.getNextBatchInvocationCount(), 2);
+        }
+    }
+
+    public void testDriverGroups()
+    {
+//        try (SplitSource source = new BufferingSplitSource(mockSource, 100)) {
+//            assertNextBatchFailure(source, 2, nextBatchCall);
+//            assertEquals(mockSource.getNextBatchInvocationCount(), 2);
+//        }
+    }
+
+    private ListenableFuture<?> invokeNextBatch(SplitSource source, int maxSize, NextBatchCall nextBatchCall)
+    {
+        ListenableFuture<?> result;
+        switch (nextBatchCall) {
+            case SINGLE_ARGUMENT:
```

Comment:
What's this? It doesn't seem to serve any purpose in this PR, so either remove it or move it to the PR that's going to need it.

Reply:
We talked about this offline. I'm avoiding structural changes to in the future commit so that that one will be easier to review. As you have realized, reviewing this file is quite involving despite it made little material change.